File: 0_3.leetcode/java/Implement_Strstr.java
public class Implement_Strstr {
    public int strStr(String haystack, String needle) {
        int m = haystack.length();
        int n = needle.length();

        if (m < n) {
            return -1;
        }

        for (int i=0;i<=(m-n);i++) {
            if (haystack.substring(i, i + n).equals(needle)) {
                return i;
            }
        }

        return -1;
    }
}

File: 0_3.leetcode/kotlin/src/main/kotlin/Implement_Strstr.kt
fun strStr(haystack: String, needle: String): Int {
    val h = haystack.length
    val n = needle.length
    var hIdx = 0
    var nIdx = 0

    while (hIdx < h && nIdx < n) {
        if (haystack[hIdx] == needle[nIdx]) {
            nIdx++
        } else {
            hIdx -= nIdx
            nIdx = 0
        }
        hIdx++
    }

    return if (nIdx == n) {
        hIdx - n
    } else {
        -1
    }
}